src/main/java/net/quarrymod/blockentity/machine/tier3/QuarryBlockEntity.java

```diff
@@ -181,98 +198,111 @@ private void refreshUpgrades() {
+        }
+    }
+
+    private boolean hasOneOfExcavationStates(ExcavationState... states) {
+        for (ExcavationState state : states) {
+            if (this.excavationState == state) {
+                return true;
             }
         }
+        return false;
     }
 
+    private void performEnergyReduction() {
+        final long euNeeded = getEnergyPerExcavation() / getTicksPerExcavation();
+        final long euAvailable = Math.min(euNeeded, getStored());
+        if (euAvailable > 0d) {
+            useEnergy(euAvailable);
+            miningEnergySpend += euAvailable;
+            setExcavationState(InProgress);
+        } else {
+            setExcavationState(ExcavationState.NoEnergyIncome);
+        }
+
+    }
+
+    @SuppressWarnings("ConstantConditions")
     private BlockPos findOrePos() {
-        final int radius = (int) Math.round(QMConfig.quarrySqrWorkRadiusByUpgradeLevel.get(rangeExtenderLevel));
+        final int currentY = getDrillTubeDepth();
+        if (currentY == pos.getY()) {
+            setExcavationState(NoOresInCurrentDepth);
+            return null;
+        }
+        final int radius = (int) Math.round(
+            QuarryMachineConfig.quarrySqrWorkRadiusByUpgradeLevel.get(rangeExtenderLevel));
 
         final BlockPos upperBlockPos = pos.add(radius, 0, radius);
         final BlockPos lowerBlockPos = pos.add(-radius, 0, -radius);
-        final ChunkPos upperChunkPos = world.getChunk(upperBlockPos).getPos();
-        final ChunkPos lowerChunkPos = world.getChunk(lowerBlockPos).getPos();
-
-        for (int y = getDrillTubeDepth(); y < pos.getY(); y++) {
-            for (int chunkX = lowerChunkPos.x; chunkX <= upperChunkPos.x; chunkX++) {
-                for (int chunkZ = lowerChunkPos.z; chunkZ <= upperChunkPos.z; chunkZ++) {
-                    if (world.isChunkLoaded(chunkX, chunkZ)) {
-                        for (int x = 0; x < 16; x++) {
-                            for (int z = 0; z < 16; z++) {
-                                int blockX = chunkX * 16 + x;
-                                int blockZ = chunkZ * 16 + z;
-
-                                if (blockX <= upperBlockPos.getX() &&
-                                    blockZ <= upperBlockPos.getZ() &&
-                                    blockX >= lowerBlockPos.getX() &&
-                                    blockZ >= lowerBlockPos.getZ()) {
-                                    BlockPos blockPos = new BlockPos(blockX, y, blockZ);
-
-                                    if (isOre(blockPos)) {
-                                        setExcavationState(ExcavationState.InProgress);
-                                        return blockPos;
-                                    }
-                                }
-                            }
-                        }
-                    }
+
+        for (int currentZ = lowerBlockPos.getZ(); currentZ < upperBlockPos.getZ(); currentZ++) {
```

Comment:
In plaats van gebruik maken van een for loop. Is ook minecraft's box api gebruiken:
Doc: https://maven.fabricmc.net/docs/yarn-21w05b+build.11/net/minecraft/util/math/Box.html
Voorbeeld: https://github.com/GabrielOlvH/Industrial-Revolution/blob/6bd774390e70c2925c446a4efe7eb699a41b2b0d/src/main/kotlin/me/steven/indrev/blockentities/farms/ChopperBlockEntity.kt#L190